# ls command


In [2]:
%%bash

# Create a test directory and navigate into it
mkdir -p test_dir/level1
cd test_dir

# Create some files and subdirectories
touch file1 file2 file3
mkdir subdir1 subdir2
cd ..
tree test_dir


mkdir: cannot create directory ‘subdir1’: File exists
mkdir: cannot create directory ‘subdir2’: File exists


test_dir
├── file1
├── file2
├── file3
├── level1
├── subdir1
└── subdir2

4 directories, 3 files


In [19]:
%%bash
cd 
echo "# Default listing of files and directories in the current directory"
ls 
echo ""
echo "# List in long format (-l) showing permissions, owner, group, size, and timestamp"
ls -l
echo ""
echo "# List with human-readable file sizes (in bytes) (-lh)"
ls -lh


# Default listing of files and directories in the current directory
rep1
testfile.txt
text_demo

# List in long format (-l) showing permissions, owner, group, size, and timestamp
total 8
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
-rw-r--r-- 1 kavis kavis   22 Oct 19 10:34 testfile.txt
drwxr-xr-x 2 kavis kavis 4096 Oct  8 10:28 text_demo

# List with human-readable file sizes (in bytes) (-lh)
total 8.0K
lrwxrwxrwx 1 kavis kavis   52 Oct  6 18:08 rep1 -> /mnt/c/Users/kavis/Upskilling/Github repository/Rep1
-rw-r--r-- 1 kavis kavis   22 Oct 19 10:34 testfile.txt
drwxr-xr-x 2 kavis kavis 4.0K Oct  8 10:28 text_demo


In [4]:
%%bash
echo "Show hidden files as well (-a), includes files starting with '.'"
ls -a
echo " "


Show hidden files as well (-a), includes files starting with '.'
.
..
.hidden_directory
.hidden_file.txt
1500x500.jpeg
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
Linux_shell_w1_d.ipynb
Linux_shell_w2_a.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
ed_editor_extended_guide.ipynb
file1
file2
level1
random_toss.py
subdir1
subdir2
test_dir
 


In [5]:
%%bash
echo "Long format with hidden files (-la), useful to see permissions of hidden files"
ls -la


Long format with hidden files (-la), useful to see permissions of hidden files
total 708
drwxrwxrwx 1 kavis kavis   4096 Oct 20 01:17 .
drwxrwxrwx 1 kavis kavis   4096 Oct 19 17:20 ..
drwxrwxrwx 1 kavis kavis   4096 Oct  8 08:38 .hidden_directory
-rwxrwxrwx 1 kavis kavis      0 Oct  8 08:38 .hidden_file.txt
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  15710 Oct 20 01:16 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavi

In [6]:

%%bash
echo "List in long format with hidden files and include inode number (-lia)"
ls -lia


List in long format with hidden files and include inode number (-lia)
total 708
 1970324837338733 drwxrwxrwx 1 kavis kavis   4096 Oct 20 01:17 .
 1688849860627019 drwxrwxrwx 1 kavis kavis   4096 Oct 19 17:20 ..
18014398509607323 drwxrwxrwx 1 kavis kavis   4096 Oct  8 08:38 .hidden_directory
23643898043815456 -rwxrwxrwx 1 kavis kavis      0 Oct  8 08:38 .hidden_file.txt
22236523160207455 -rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
 1688849860628098 -rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
38562071809434349 -rwxrwxrwx 1 kavis kavis  15710 Oct 20 01:16 Linux_shell_w1_a.ipynb
 5910974511104067 -rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
 9007199254921272 -rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
 5629499534414068 -rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
38280596832693534 -rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
 1970324837338746 -rwxrwxr

In [7]:
%%bash
echo " "
echo "Show the inode numbers (-i) of files and directories in the current directory"
ls -i


 
Show the inode numbers (-i) of files and directories in the current directory
22236523160207455 1500x500.jpeg
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 5629499534414068 Linux_shell_w1_d.ipynb
38280596832693534 Linux_shell_w2_a.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb
21673573206756103 ed_editor_extended_guide.ipynb
 2533274791046055 file1
 1970324837624745 file2
 5910974511298353 level1
21673573206783692 random_toss.py
 2533274791046077 subdir1
 2533274791046079 subdir2
45035996273708736 test_dir


In [4]:
%%bash
pwd
ls -d

#-d: This flag prevents ls from listing the contents of the directories

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
.


The -d option with the ls command in Linux tells ls to display directories as items themselves rather than listing their contents

In [36]:
%%bash
# Use -d to show information only about the directory itself, not its contents
ls -ld test_dir/


drwxrwxrwx 1 kavis kavis 4096 Oct 20 01:17 test_dir/


In [5]:
%%bash
echo "List in long format of only directories, showing details of directories themselves (-ld)"
ls -ld */
echo " "

#ls -d */: The */ syntax is used to list directories only (with a trailing slash indicating directories), 
# ignoring files in the current directory.
# -d: This flag prevents ls from listing the contents of the directories
# */: This pattern matches directories only

echo "Show inode numbers along with hidden files (-ia)"
ls -ia

List in long format of only directories, showing details of directories themselves (-ld)
drwxrwxrwx 1 kavis kavis 4096 Oct 26 14:30 empty_directory/
drwxrwxrwx 1 kavis kavis 4096 Oct 26 14:44 test_dir/
 
Show inode numbers along with hidden files (-ia)
 1970324837338733 .
 1688849860627019 ..
18014398509607323 .hidden_directory
23643898043815456 .hidden_file.txt
22236523160207455 1500x500.jpeg
20547673299915561 Lab_questions.ipynb
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 5629499534414068 Linux_shell_w1_d.ipynb
38280596832693534 Linux_shell_w2_a.ipynb
13792273858909618 Linux_shell_w2_b.ipynb
 4503599627517758 Linux_shell_w2_c.ipynb
 7599824371351807 Linux_shell_w2_d.ipynb
12666373952166606 Linux_shell_w3_a.ipynb
 6192449487773015 Linux_shell_w3_b.ipynb
11540474045231995 Linux_shell_w3_c.ipynb
18295873486235463 Linux_shell_w3_d.ipynb
10696049115054630 Linux

In [18]:
%%bash
echo "Show file sizes in blocks (-s)"
ls -s
echo " "


Show file sizes in blocks (-s)
total 708
164 1500x500.jpeg
 24 Linux_shell_setup_windows.ipynb
 16 Linux_shell_w1_a.ipynb
 48 Linux_shell_w1_b.ipynb
 40 Linux_shell_w1_c.ipynb
100 Linux_shell_w1_d.ipynb
 40 Linux_shell_w2_a.ipynb
  8 SC_Week0_Unit0.ipynb
228 SC_Week0_Unit1.ipynb
 16 SC_Week0_Unit2.ipynb
  4 SC_Week0_Unit4.ipynb
  8 SC_Week0_Unit5.ipynb
 12 ed_editor_extended_guide.ipynb
  0 file1
  0 file2
  0 level1
  0 random_toss.py
  0 subdir1
  0 subdir2
  0 test_dir
 


When you use the command ls -s in Linux or Unix-like systems, it shows the size of files in blocks instead of the usual bytes or kilobytes. 

What is a "Block"?
Think of your computer's storage (like a hard drive or SSD) as a big grid made up of small sections called blocks.
Each block is a fixed-size unit of storage, usually 4 kilobytes (KB) on modern systems, though it can vary.
When you store a file, it uses one or more of these blocks, depending on the size of the file. Even if a file is very small, like 1 byte, it will still take up at least one block (e.g., 4 KB).

In [20]:
%%bash
echo "Recursive listing, show contents of all directories and subdirectories (-R)"
ls -R


Recursive listing, show contents of all directories and subdirectories (-R)
.:
1500x500.jpeg
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
Linux_shell_w1_d.ipynb
Linux_shell_w2_a.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
ed_editor_extended_guide.ipynb
file1
file2
level1
random_toss.py
subdir1
subdir2
test_dir

./level1:
file3
level2

./level1/level2:
file4

./subdir1:

./subdir2:

./test_dir:
file1
file2
file3
level1
subdir1
subdir2

./test_dir/level1:

./test_dir/subdir1:

./test_dir/subdir2:


In [21]:

%%bash
echo "Recursive listing in long format (-lR)"
ls -lR


Recursive listing in long format (-lR)
.:
total 708
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  15824 Oct 20 01:28 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
-rwxrwxrwx 1 kavis kavis   8765 Oct 19 11:33 ed_editor_extended_guide.ipynb
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file1
-rwxrwxrwx 1 ka

In [6]:
%%bash
# Show files sorted by size (-lS) to get largest files first
ls -lS



total 1468
-rwxrwxrwx 1 kavis kavis 345643 Oct 23 09:40 Linux_shell_w2_c.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 19:51 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 21:00 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis 112827 Oct 20 11:12 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  84362 Oct 23 14:37 Linux_shell_w2_d.ipynb
-rwxrwxrwx 1 kavis kavis  67233 Oct 20 07:38 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  52150 Oct 20 10:28 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis  43854 Oct 23 17:45 Linux_shell_w3_e.ipynb
-rwxrwxrwx 1 kavis kavis  40034 Oct 23 17:32 Linux_shell_w3_d.ipynb
-rwxrwxrwx 1 kavis kavis  39684 Oct 23 17:35 Linux_shell_w3_c.ipynb
-rwxrwxrwx 1 kavis kavis  37083 Oct 23 16:43 Linux_shell_w3_b.ipynb
-rwxrwxrwx 1 kavis kavis  34663 Oct 23 18:08 Linux_shell_w4b.ipynb
-rwxrwxrwx 1 kavis kavis  33192 Oct 23 09:38 Linux_shell_w2_b.ipynb
-rwxrwxrwx 1 kavis kavis  28780 Oct 20 12:56 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis  25450 Oct 23 16:27 Linu

In [25]:
%%bash
# List files sorted by modification time, newest first (-lt)
ls -lt


total 708
-rwxrwxrwx 1 kavis kavis  15824 Oct 20 01:28 Linux_shell_w1_a.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct 20 01:17 test_dir
-rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis   8765 Oct 19 11:33 ed_editor_extended_guide.ipynb
-rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 subdir2
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 subdir1
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 level1
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file2
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file1
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipyn

In [26]:

%%bash
# List files in reverse order by modification time (-ltr)
ls -ltr


total 708
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis    199 Oct  6 19:47 random_toss.py
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file1
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file2
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 level1
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 subdir1
drwxrwxrwx 1 kavis kavis   4096 Oct  9 03:07 subdir2
-rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb


In [27]:
%%bash
# Combine -i and -l to show inode numbers along with detailed view (-il)
ls -il


total 708
22236523160207455 -rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
 1688849860628098 -rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
38562071809434349 -rwxrwxrwx 1 kavis kavis  15824 Oct 20 01:28 Linux_shell_w1_a.ipynb
 5910974511104067 -rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
 9007199254921272 -rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
 5629499534414068 -rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
38280596832693534 -rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
 1970324837338746 -rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
 1688849860628082 -rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
 1688849860628085 -rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
 1688849860628079 -rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
 1688849860628086 -rwxrwxrwx 1 kavis kavis   4159 Oc

In [28]:

%%bash
# Combine -i, -a, and -l to show inode numbers, hidden files, and detailed view (-ial)
ls -ial


total 708
 1970324837338733 drwxrwxrwx 1 kavis kavis   4096 Oct 20 01:17 .
 1688849860627019 drwxrwxrwx 1 kavis kavis   4096 Oct 19 17:20 ..
18014398509607323 drwxrwxrwx 1 kavis kavis   4096 Oct  8 08:38 .hidden_directory
23643898043815456 -rwxrwxrwx 1 kavis kavis      0 Oct  8 08:38 .hidden_file.txt
22236523160207455 -rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
 1688849860628098 -rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
38562071809434349 -rwxrwxrwx 1 kavis kavis  15824 Oct 20 01:28 Linux_shell_w1_a.ipynb
 5910974511104067 -rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
 9007199254921272 -rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
 5629499534414068 -rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
38280596832693534 -rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
 1970324837338746 -rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
 16888498606

In [33]:
%%bash

# Navigate into the "level1" directory and list contents
cd test_dir/level1
pwd
ls -l

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell/test_dir/level1
total 0


In [34]:
%%bash
# Go back to parent directory and list contents recursively with details (-lR)
ls -lR


.:
total 708
-rwxrwxrwx 1 kavis kavis 166075 Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis  23480 Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  15824 Oct 20 01:28 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis  48873 Oct  9 03:18 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  40647 Oct  9 03:56 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis 101837 Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  40244 Oct 19 17:21 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis   8173 Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 230730 Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis  16186 Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis   3531 Oct  5 14:22 SC_Week0_Unit4.ipynb
-rwxrwxrwx 1 kavis kavis   4159 Oct  5 14:21 SC_Week0_Unit5.ipynb
-rwxrwxrwx 1 kavis kavis   8765 Oct 19 11:33 ed_editor_extended_guide.ipynb
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file1
-rwxrwxrwx 1 kavis kavis      0 Oct  9 03:07 file2
drw

In [15]:
%%bash
# This option shows all files, including hidden ones (files starting with .), 
# but it excludes the special directories . (current directory) and .. (parent directory).
ls -A


.hidden_directory
.hidden_file.txt
1500x500.jpeg
Lab_questions.ipynb
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
Linux_shell_w1_d.ipynb
Linux_shell_w2_a.ipynb
Linux_shell_w2_b.ipynb
Linux_shell_w2_c.ipynb
Linux_shell_w2_d.ipynb
Linux_shell_w3_a.ipynb
Linux_shell_w3_b.ipynb
Linux_shell_w3_c.ipynb
Linux_shell_w3_d.ipynb
Linux_shell_w3_e.ipynb
Linux_shell_w4a.ipynb
Linux_shell_w4b.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
anotherfile.txt
block_size_file.txt
empty_directory
myfile.txt
random_toss.py
random_toss.py.ipynb
test.bashrc
test_dir
testfile.txt


In [16]:
%%bash
ls -d

.


In [13]:
%%bash
ls -Ad

.


In [17]:
%%bash
ls *

1500x500.jpeg
Lab_questions.ipynb
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
Linux_shell_w1_d.ipynb
Linux_shell_w2_a.ipynb
Linux_shell_w2_b.ipynb
Linux_shell_w2_c.ipynb
Linux_shell_w2_d.ipynb
Linux_shell_w3_a.ipynb
Linux_shell_w3_b.ipynb
Linux_shell_w3_c.ipynb
Linux_shell_w3_d.ipynb
Linux_shell_w3_e.ipynb
Linux_shell_w4a.ipynb
Linux_shell_w4b.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
anotherfile.txt
block_size_file.txt
myfile.txt
random_toss.py
random_toss.py.ipynb
test.bashrc
testfile.txt

empty_directory:
another_empty_directory

test_dir:
file1
file2
file3
level1
subdir1
subdir2


In [ ]:
%%bash

ls -A *

# ls -A *: Lists files (including hidden ones) in the current directory as well as the contents of all subdirectories.
# The * is a wildcard character that expands to match all files and directories in the current directory.

1500x500.jpeg
Lab_questions.ipynb
Linux_shell_setup_windows.ipynb
Linux_shell_w1_a.ipynb
Linux_shell_w1_b.ipynb
Linux_shell_w1_c.ipynb
Linux_shell_w1_d.ipynb
Linux_shell_w2_a.ipynb
Linux_shell_w2_b.ipynb
Linux_shell_w2_c.ipynb
Linux_shell_w2_d.ipynb
Linux_shell_w3_a.ipynb
Linux_shell_w3_b.ipynb
Linux_shell_w3_c.ipynb
Linux_shell_w3_d.ipynb
Linux_shell_w3_e.ipynb
Linux_shell_w4a.ipynb
Linux_shell_w4b.ipynb
SC_Week0_Unit0.ipynb
SC_Week0_Unit1.ipynb
SC_Week0_Unit2.ipynb
SC_Week0_Unit4.ipynb
SC_Week0_Unit5.ipynb
anotherfile.txt
block_size_file.txt
myfile.txt
random_toss.py
random_toss.py.ipynb
test.bashrc
testfile.txt

empty_directory:
another_empty_directory

test_dir:
file1
file2
file3
level1
subdir1
subdir2


Use ls -R for a full directory tree view 

ls -A * when you want a detailed look only at the current directory's contents and each top-level folder within it.

In [18]:
%%bash
# List all files including hidden ones and inodes, but just at the top-level directory (-iAd)
ls -iAd *


22236523160207455 1500x500.jpeg
20547673299915561 Lab_questions.ipynb
 1688849860628098 Linux_shell_setup_windows.ipynb
38562071809434349 Linux_shell_w1_a.ipynb
 5910974511104067 Linux_shell_w1_b.ipynb
 9007199254921272 Linux_shell_w1_c.ipynb
 5629499534414068 Linux_shell_w1_d.ipynb
38280596832693534 Linux_shell_w2_a.ipynb
13792273858909618 Linux_shell_w2_b.ipynb
 4503599627517758 Linux_shell_w2_c.ipynb
 7599824371351807 Linux_shell_w2_d.ipynb
12666373952166606 Linux_shell_w3_a.ipynb
 6192449487773015 Linux_shell_w3_b.ipynb
11540474045231995 Linux_shell_w3_c.ipynb
18295873486235463 Linux_shell_w3_d.ipynb
10696049115054630 Linux_shell_w3_e.ipynb
27303072740954676 Linux_shell_w4a.ipynb
63331869759912754 Linux_shell_w4b.ipynb
 1970324837338746 SC_Week0_Unit0.ipynb
 1688849860628082 SC_Week0_Unit1.ipynb
 1688849860628085 SC_Week0_Unit2.ipynb
 1688849860628079 SC_Week0_Unit4.ipynb
 1688849860628086 SC_Week0_Unit5.ipynb
16325548649361481 anotherfile.txt
56294995342191540 block_size_file.txt


ls -iAd *: Lists content in the current directory, treating each directory as an item without showing its contents.
ls -iA: Lists contents in the current directory, without expanding or treating directories as single items.

The -d option in ls -iAd * prevents directory expansion, while ls -iA * shows contents for each directory matched by *

In [45]:
%%bash
echo " "
# List files with detailed view and human-readable sizes, sort by size (-lhS)
ls -lhS


 
total 708K
-rwxrwxrwx 1 kavis kavis 226K Oct  5 14:21 SC_Week0_Unit1.ipynb
-rwxrwxrwx 1 kavis kavis 163K Oct  6 15:30 1500x500.jpeg
-rwxrwxrwx 1 kavis kavis 100K Oct  9 03:52 Linux_shell_w1_d.ipynb
-rwxrwxrwx 1 kavis kavis  48K Oct  9 03:18 Linux_shell_w1_b.ipynb
-rwxrwxrwx 1 kavis kavis  40K Oct  9 03:56 Linux_shell_w1_c.ipynb
-rwxrwxrwx 1 kavis kavis  40K Oct 19 17:21 Linux_shell_w2_a.ipynb
-rwxrwxrwx 1 kavis kavis  23K Oct  6 14:32 Linux_shell_setup_windows.ipynb
-rwxrwxrwx 1 kavis kavis  16K Oct  5 14:58 SC_Week0_Unit2.ipynb
-rwxrwxrwx 1 kavis kavis  16K Oct 20 01:28 Linux_shell_w1_a.ipynb
-rwxrwxrwx 1 kavis kavis 8.6K Oct 19 11:33 ed_editor_extended_guide.ipynb
-rwxrwxrwx 1 kavis kavis 8.0K Oct  6 19:54 SC_Week0_Unit0.ipynb
-rwxrwxrwx 1 kavis kavis 4.1K Oct  5 14:21 SC_Week0_Unit5.ipynb
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 level1
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 subdir1
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 subdir2
drwxrwxrwx 1 kavis kavis 4.0K Oct 20 01:17 

In [46]:
%%bash
# List directories themselves in detailed view with inode numbers (-ild)
ls -ild */


 5910974511298353 drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 level1/
 2533274791046077 drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 subdir1/
 2533274791046079 drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 subdir2/
45035996273708736 drwxrwxrwx 1 kavis kavis 4096 Oct 20 01:17 test_dir/


In [20]:
%%bash
# Create a test directory structure with nested subdirectories and files
mkdir -p level1/level2
touch file1 file2 level1/file3 level1/level2/file4
mkdir subdir1 subdir2
cd ..
tree Linux_shell


Linux_shell
├── 1500x500.jpeg
├── Lab_questions.ipynb
├── Linux_shell_setup_windows.ipynb
├── Linux_shell_w1_a.ipynb
├── Linux_shell_w1_b.ipynb
├── Linux_shell_w1_c.ipynb
├── Linux_shell_w1_d.ipynb
├── Linux_shell_w2_a.ipynb
├── Linux_shell_w2_b.ipynb
├── Linux_shell_w2_c.ipynb
├── Linux_shell_w2_d.ipynb
├── Linux_shell_w3_a.ipynb
├── Linux_shell_w3_b.ipynb
├── Linux_shell_w3_c.ipynb
├── Linux_shell_w3_d.ipynb
├── Linux_shell_w3_e.ipynb
├── Linux_shell_w4a.ipynb
├── Linux_shell_w4b.ipynb
├── SC_Week0_Unit0.ipynb
├── SC_Week0_Unit1.ipynb
├── SC_Week0_Unit2.ipynb
├── SC_Week0_Unit4.ipynb
├── SC_Week0_Unit5.ipynb
├── anotherfile.txt
├── block_size_file.txt
├── empty_directory
│   └── another_empty_directory
├── file1
├── file2
├── level1
│   ├── file3
│   └── level2
│       └── file4
├── myfile.txt
├── random_toss.py
├── random_toss.py.ipynb
├── subdir1
├── subdir2
├── test.bashrc
└── testfile.txt

7 directories, 34 files


In [48]:
%%bash
# List contents of a specific directory using it as an argument
ls level1


file3
level2


In [50]:
%%bash
# List contents of multiple directories (subdir1, level1) using directory names as arguments
ls subdir1 level1

echo " "
# List contents of multiple directories in long format (-l)
ls -l subdir1 level1


level1:
file3
level2

subdir1:
 


level1:
total 0
-rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file3
drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 level2

subdir1:
total 0


In [52]:
%%bash
# List the contents of level1 recursively (-R), showing all files in subdirectories
ls -R level1



level1:
file3
level2

level1/level2:
file4


In [53]:

%%bash
# List detailed view of the level1 directory itself without showing its contents (-ld)
ls -ld level1


drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 level1


In [54]:
%%bash
# Using -ld to show information about the directory "level1/level2" without listing its contents
ls -ld level1/level2


drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 level1/level2


In [55]:
%%bash
# List with detailed view and human-readable sizes (-lh) for multiple directories
ls -lh level1 subdir1


level1:
total 0
-rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file3
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 level2

subdir1:
total 0


In [57]:
%%bash
# Recursive listing for multiple directories (-lR)
ls -lR level1 subdir2



level1:
total 0
-rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file3


drwxrwxrwx 1 kavis kavis 4096 Oct  9 03:07 level2

level1/level2:
total 0
-rwxrwxrwx 1 kavis kavis 0 Oct 20 01:51 file4

subdir2:
total 0


In [58]:
%%bash
# Combine options to show all directories and subdirectories recursively with human-readable sizes (-alhR)
ls -alhR level1 subdir1 subdir2


level1:
total 0
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 .
drwxrwxrwx 1 kavis kavis 4.0K Oct 20 01:17 ..
-rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file3
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 level2

level1/level2:
total 0
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 .
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 ..
-rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file4

subdir1:
total 0
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 .
drwxrwxrwx 1 kavis kavis 4.0K Oct 20 01:17 ..

subdir2:
total 0
drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 .
drwxrwxrwx 1 kavis kavis 4.0K Oct 20 01:17 ..


In [59]:

%%bash
# Use a combination of -i (inode numbers) and directory arguments
ls -i level1 level1/level2



level1:
2533274791046059 file3
2533274791046036 level2

level1/level2:
2533274791046069 file4


In [60]:
%%bash
# List inode numbers and long format with human-readable sizes for multiple directories (-ilh)
ls -ilh level1 subdir1 subdir2


level1:
total 0
2533274791046059 -rwxrwxrwx 1 kavis kavis    0 Oct 20 01:51 file3
2533274791046036 drwxrwxrwx 1 kavis kavis 4.0K Oct  9 03:07 level2

subdir1:
total 0

subdir2:
total 0


In [22]:
%%bash
# Clean up: remove the created test directory
rm -r level1 subdir1 subdir2


rm: cannot remove 'level1': No such file or directory
rm: cannot remove 'subdir1': No such file or directory
rm: cannot remove 'subdir2': No such file or directory


CalledProcessError: Command 'b'# Clean up: remove the created test directory\nrm -r level1 subdir1 subdir2\n'' returned non-zero exit status 1.

In [ ]:
'''
%%bash

cd
pwd

# Create a sample directory structure
mkdir -p sample_dir/subdir
cd sample_dir
touch file1 file2 .hidden_file subdir/file3

# Short form of options:
ls -a          # Show all files, including hidden files
ls -l          # Show long format listing
ls -la         # Combine -l and -a to show hidden files in long format
ls -lh         # Long format with human-readable file sizes
ls -R          # Recursively list all directories and their contents
ls -d */       # Show directory names only
ls -i          # Show inode numbers
ls -r          # Reverse the order of the listing
ls -S          # Sort by file size
ls -t          # Sort by modification time

# Long form of options:
ls --all           # Equivalent to -a, show all files including hidden ones
ls --long          # Equivalent to -l, show long format listing
ls --all --long    # Equivalent to -la, show hidden files in long format
ls --human-readable --long  # Long format with human-readable sizes
ls --recursive     # Equivalent to -R, show directories and their contents recursively
ls --directory */  # Equivalent to -d, show directory names only
ls --inode         # Equivalent to -i, show inode numbers
ls --reverse       # Equivalent to -r, show files in reverse order
ls --sort=size     # Equivalent to -S, sort by file size
ls --time          # Equivalent to -t, sort by modification time

# Clean up: remove the created test directory
cd ..
rm -r sample_dir
'''


Key Points to Remember:
Short Form: Typically a single character preceded by a hyphen (-), e.g., -l, -a, -R.

Long Form: A descriptive word preceded by two hyphens (--), e.g., --all, --recursive, --long.

Short and long forms can be used interchangeably. For example, -l is equivalent to --long.

Multiple short options can be combined (e.g., ls -laR), but long options must be written separately (e.g., ls --long --all --recursive).

Use man ls or ls --help to see all available short and long forms for ls.

Short Form vs. Long Form of ls Options

Short Form	Long Form	Description
-a	          --all	    Show all files, including hidden files (those starting with .).

-A	       --almost-all	Show all files except . (current directory) and .. (parent).

-d	     --directory	List directory names only, not their contents.

-h	--human-readable	Show sizes in human-readable format (KB, MB, etc.).

-l	          --long	Use a long listing format to show detailed information.
-R	     --recursive	List subdirectories recursively.

-S	     --sort=size	Sort files by size (largest first).

-t	          --time	Sort by modification time (newest first).

-r	--reverse	Reverse the order of sorting.

-i	--inode	Show inode numbers.

-s	--size	Display file sizes in blocks.

-u	--access-time	Sort by last access time instead of modification time.

-c	--time=ctime	Sort by change time (status change time).

-G	--no-group	Do not show group information in the long format.

-X	--sort=extension	Sort alphabetically by file extension.

-1	--format=single-column	List one file per line.

This sequence includes a broad range of options for ls:

-l: Long format.
-h: Human-readable sizes.
-a: Include hidden files.
-i: Show inode numbers.
-d: Show directory information only.
-R: Recursive listing.
-S: Sort by size.
-t: Sort by modification time.
-r: Reverse the order.
-ia: Include inode numbers and hidden files.

Explanation of ls with Directories as Arguments:


ls <directory_name>: Lists the contents of the specified directory, not the current working directory.
Example: ls level1 lists files inside level1 directory.


ls <dir1> <dir2>: Lists contents of multiple directories.
Example: ls level1 subdir1 lists files in both level1 and subdir1.


ls -ld <directory>: Lists information about the directory itself, not its contents.
Example: ls -ld level1 shows details (like permissions, owner, etc.) of the directory level1 itself.


Recursive Listing (ls -R <directory>): Displays all files and subdirectories within the specified directory recursively.
Example: ls -R level1 shows all files in level1 and its subdirectories.


Combine Options with Directory Arguments:
ls -alhR level1 subdir1: Shows all hidden files in a long format, human-readable sizes, and recursively lists contents of level1 and subdir1.

In [68]:
%%bash
pwd
cd /mnt/c
ls -l
clear

/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell


ls: DumpStack.log: Permission denied
ls: cannot access 'DumpStack.log.tmp': Permission denied
ls: cannot access 'hiberfil.sys': Permission denied
ls: cannot access 'pagefile.sys': Permission denied
ls: PerfLogs: Permission denied
ls: Recovery: Permission denied
ls: cannot access 'swapfile.sys': Permission denied
ls: 'System Volume Information': Permission denied


total 16
drwxrwxrwx 1 kavis kavis  4096 Oct 20  2022 $Recycle.Bin
-r-xr-xr-x 1 kavis kavis    40 Oct 20  2022 2CA0AE717CAA
lrwxrwxrwx 1 kavis kavis    12 Oct 10  2022 Documents and Settings -> /mnt/c/Users
drwxrwxrwx 1 kavis kavis  4096 Jul 30  2022 Drivers
---------- 1 kavis kavis 12288 Jul 12 03:44 DumpStack.log
-????????? ? ?     ?         ?            ? DumpStack.log.tmp
drwxrwxrwx 1 kavis kavis  4096 May  4 08:13 OneDriveTemp
d--x--x--x 1 kavis kavis  4096 May  7  2022 PerfLogs
dr-xr-xr-x 1 kavis kavis  4096 Oct  6 03:43 Program Files
dr-xr-xr-x 1 kavis kavis  4096 Jun 11 03:17 Program Files (x86)
drwxrwxrwx 1 kavis kavis  4096 Sep 19 16:32 ProgramData
d--x--x--x 1 kavis kavis  4096 Aug 10  2023 Recovery
-r-xr-xr-x 1 kavis kavis  1505 Dec 19  2023 Setup.log
d--x--x--x 1 kavis kavis  4096 Oct 19 13:01 System Volume Information
drwxrwxrwx 1 kavis kavis  4096 Dec  3  2023 Tenorshare
dr-xr-xr-x 1 kavis kavis  4096 Jul 10 09:50 Users
dr-xr-xr-x 1 kavis kavis  4096 Oct 12 11:36 Windows



#### Structure of Each Line given by ls -l:
Structure will change slightly depending on VS code is run in wsl or not

1. **Permissions**: (`drwxr-xr-x`)
   - This shows the file permissions for the owner, group, and others.
   - first letter -  `d` indicates a **directory** (if it were a file, it would be `-`).

                     'l' indicates symbolic link (shortcut)

                     '-' indicates file 
                     
                     'c' character device (terminal)
                     
                     'b' block devices (hard discs)
   - next three for owner - `rwx` means **read, write, and execute** permissions for the owner.
   - next three for group - `r-x` means **read and execute** permissions for the group.
   - next three for others- `r-x` means **read and execute** permissions for others.

   rwx (7)
   rw- (6)
   r-x (5)
   r-- (4)
   --x (1)

2. **Links**: (`1`)
   - The number of links or directories inside this directory. For a directory, this number (1) represents how many subdirectories, parent directories, or other references link to it. A hard link is like an alias or shortcut that points to the same file or directory in the filesystem. 

3. **Owner**: (`kavis`)
   - The username of the owner of the file or directory.

4. **Group**: (`197609`)
   - The group ID associated with the file.

5. **Size**: (`0`, `40`, `12288`, etc.)
   - The size of the file in bytes. For directories, it’s usually `0` since it doesn’t directly reflect the size of contents.

6. **Timestamp**: (`Oct 6 14:21`)
   - The date and time when the file or directory was last modified.

7. **File/Directory Name**: (`$Recycle.Bin`, `Documents and Settings`, etc.)
   - The name of the file or directory.

#### Listing Explanation:

**`Documents and Settings -> /c/Users`**:
   - This is a **symbolic link** (shortcut) that points to `/c/Users`. It means that trying to access `Documents and Settings` will actually redirect you to the `/c/Users` directory.

**`Drivers`**:
   - Likely a folder containing drivers for the system. Drivers are software that help the operating system communicate with hardware devices.

**`Program Files`** and **`Program Files (x86)`**:
   - Standard directories in Windows for installing 64-bit and 32-bit applications, respectively.

**`ProgramData`**:
   - A directory used to store application data that is not specific to a user. It’s shared among all users of the system.

**`Users`**:
    - This is where user profiles are stored on Windows, such as `Desktop`, `Documents`, `Downloads`, and so on for each user.

**`Windows`**:
    - The main directory containing the Windows operating system files.

**`hiberfil.sys`**:
    - A **system file** used by Windows to store the system state when it goes into **hibernation**. This file is often large and contains everything stored in memory (RAM).

In [70]:
%%bash
pwd
cd /
ls -im  #ls -i shows the inode number associated with each file or directory, which helps identify them uniquely


/mnt/c/Users/kavis/Upskilling/Github repository/Rep1/Linux_shell
12 bin, 8193 bin.usr-is-merged, 8194 boot, 1 dev, 16386 etc, 24577 home, 5 init,
1395 lib, 24578 lib.usr-is-merged, 1396 lib64, 11 lost+found, 32769 media,
32770 mnt, 40961 opt, 1 proc, 49153 root, 1 run, 1424 sbin,
57345 sbin.usr-is-merged, 57346 snap, 65537 srv, 1 sys, 73729 tmp, 73730 usr,
33554433 var


In [71]:

%%bash
cd /
ls -l

total 2284
lrwxrwxrwx   1 root root       7 Apr 22 13:08 bin -> usr/bin
drwxr-xr-x   2 root root    4096 Feb 26  2024 bin.usr-is-merged
drwxr-xr-x   2 root root    4096 Apr 22 13:08 boot
drwxr-xr-x  16 root root    3560 Oct 19 09:35 dev
drwxr-xr-x  92 root root    4096 Oct 20 01:20 etc
drwxr-xr-x   3 root root    4096 Oct  5 12:54 home
-rwxrwxrwx   1 root root 2260248 Sep 27 00:55 init
lrwxrwxrwx   1 root root       7 Apr 22 13:08 lib -> usr/lib
drwxr-xr-x   2 root root    4096 Apr  8  2024 lib.usr-is-merged
lrwxrwxrwx   1 root root       9 Apr 22 13:08 lib64 -> usr/lib64
drwx------   2 root root   16384 Oct  5 12:49 lost+found
drwxr-xr-x   2 root root    4096 Sep 27 20:13 media
drwxr-xr-x   5 root root    4096 Oct  5 12:49 mnt
drwxr-xr-x   2 root root    4096 Sep 27 20:13 opt
dr-xr-xr-x 311 root root       0 Oct 19 09:35 proc
drwx------   4 root root    4096 Oct  5 12:49 root
drwxr-xr-x  18 root root     540 Oct 19 09:35 run
lrwxrwxrwx   1 root root       8 Apr 22 13:08 sbin -> usr/sb

/bin -       essential command binaries - files in machine language for processor

/usr/bin -   User commands

/usr/lib -   Libraries